### PR TITLE
Add ReagentWhitelist to Injector component and system

### PR DESCRIPTION
**Pull request by blueDev2**
## About the PR

Added a new field to injectorComponent called ReagentWhitelist. If the ReagentWhitelist is not null, the injector component can only draw reagents in the ReagentWhitelist.

## Why / Balance

The main purpose of this PR is for making an entity downstream. However, since it changes a file from this codebase and is flexible enough for other purposes, I decided to see if there's any demand for this field here.

## Technical details
During a draw action, check if there is a ReagentWhitelist. If there is one, remove non-whitelisted reagents from the draw-target-solution and put that aside. Once the actual draw has completed, re-add the removed reagents back into the draw-target-solution.

## Media

https://github.com/space-wizards/space-station-14/assets/89804215/05c5f27e-f9d1-4f66-879d-eea23b2208e5
- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase
The demo uses the following entity prototype: 
- type: entity
  parent: BaseSyringe
  name: NutirmentOnlySyringe
  id: nutirmentOnlySyringe
  components:
  - type: Injector
    transferAmount: 15
    toggleState: Draw
    reagentWhitelist:
      - Nutriment
  - type: Tag
    tags:
    - Syringe

This PR will have no affect on gameplay, it is just to make injectorComponents more flexible

### CODE REVIEW

File: Content.Server/Chemistry/EntitySystems/InjectorSystem.cs [hunk 1]
```diff
@@ -330,8 +330,21 @@ private void AfterDraw(Entity<InjectorComponent> injector, EntityUid target)
             return;
         }
 
+        var applicableTargetSolution = targetSolution.Comp.Solution;
+        // If a whitelist exists, remove all non-whitelisted reagents from the target solution temporarily
+        Solution temporarilyRemovedSolution = new();
```
Comment (moonheart08) [inline Content.Server/Chemistry/EntitySystems/InjectorSystem.cs:335]:
```suggestion
        var temporarilyRemovedSolution = new Solution();
```

File: Content.Server/Chemistry/EntitySystems/InjectorSystem.cs [hunk 2]
```diff
@@ -330,8 +330,21 @@ private void AfterDraw(Entity<InjectorComponent> injector, EntityUid target)
...
+        if (injector.Comp.ReagentWhitelist is { } reagentWhitelist)
+        {
+            string[] reagentPrototypeWhitelistArray = new string[reagentWhitelist.Count];
+            for (var i = 0; i < reagentWhitelist.Count; ++i)
+            {
+                reagentPrototypeWhitelistArray[i] = reagentWhitelist[i];
+            }
```
Comment (moonheart08) [inline Content.Server/Chemistry/EntitySystems/InjectorSystem.cs:342]:
```suggestion
            var reagentPrototypeWhitelistArray = reagentWhitelist.Select(x => (string)x).ToArray();
```

**Review by moonheart08 — CHANGES_REQUESTED** [review]

File: Content.Server/Chemistry/EntitySystems/InjectorSystem.cs (hunk 2 above)
Comment (moonheart08) [reply Content.Server/Chemistry/EntitySystems/InjectorSystem.cs:342]: note: If not for the fact C# does require the name to be different, I would be recommending the name be shortened to not include it's type (as it's redundant information), as well.

File: Content.Server/Chemistry/EntitySystems/InjectorSystem.cs [hunk 3]
```diff
@@ -330,8 +331,17 @@ private void AfterDraw(Entity<InjectorComponent> injector, EntityUid target)
             return;
         }
 
+        var applicableTargetSolution = targetSolution.Comp.Solution;
+        // If a whitelist exists, remove all non-whitelisted reagents from the target solution temporarily
+        var temporarilyRemovedSolution = new Solution();
+        if (injector.Comp.ReagentWhitelist is { } reagentWhitelist)
+        {
+            var reagentPrototypeWhitelistArray = reagentWhitelist.Select(x => (string) x).ToArray();
+            temporarilyRemovedSolution = applicableTargetSolution.SplitSolutionWithout(applicableTargetSolution.Volume, reagentPrototypeWhitelistArray);
```
Comment (metalgearsloth) [inline Content.Server/Chemistry/EntitySystems/InjectorSystem.cs:340]: Can you do this without the LINQ as that will allocate a lot.

Comment (blueDev2) [reply Content.Server/Chemistry/EntitySystems/InjectorSystem.cs:340]: Does LINQ just suck and I should never use functions from it?

Comment (TheShuEd) [reply Content.Server/Chemistry/EntitySystems/InjectorSystem.cs]: > Does LINQ just suck and I should never use functions from it?

yes

File: Content.Server/Chemistry/EntitySystems/InjectorSystem.cs [hunk 4]
```diff
@@ -1,3 +1,4 @@
+using System.Linq;
```
Comment (TheShuEd) [inline Content.Server/Chemistry/EntitySystems/InjectorSystem.cs:1]:
```suggestion
```
scary linq

**Pull request merged**